stl/inc/chrono

```diff
@@ -619,27 +671,1275 @@ namespace chrono {
+    inline constexpr month October{10};
+    inline constexpr month November{11};
+    inline constexpr month December{12};
+
+    // Returns the number of fractional digits of _Num / _Den in the range [0, 18].
+    // If it can't be represented 6 is returned.
+    // Example: _Fractional_width(1, 8) would return 3 for 0.125.
+    constexpr unsigned _Fractional_width(intmax_t _Num, const intmax_t _Den) {
+        _STL_ASSERT(_Den > 0, "Denominator can't be less than 1.");
+        unsigned _Result = 0;
+        // doing everything in for > using a while loop
+        for (; _Num % _Den != 0 && _Result < 19; _Num = _Num % _Den * 10, ++_Result) {
+        }
+        return _Result == 19 ? 6 : _Result;
+    }
+
+    constexpr intmax_t _Pow10(const int _Exp) {
+        intmax_t _Result = 1;
+        for (int _Ix = 0; _Ix < _Exp; ++_Ix) {
+            _Result *= 10;
+        }
+        return _Result;
+    }
+
+    template <class _Duration>
+    requires _Is_duration_v<_Duration> class hh_mm_ss {
+    public:
+        static constexpr unsigned fractional_width = _Fractional_width(_Duration::period::num, _Duration::period::den);
+        using precision =
+            duration<common_type_t<typename _Duration::rep, seconds::rep>, ratio<1, _Pow10(fractional_width)>>;
+
+        constexpr hh_mm_ss() noexcept : hh_mm_ss{_Duration::zero()} {}
+        // clang-format off
+        constexpr explicit hh_mm_ss(_Duration _Dur)
+            : _Is_neg{_Dur < _Duration::zero()},
+              _Hours{_CHRONO duration_cast<_CHRONO hours>(_CHRONO abs(_Dur))},
+              _Mins{_CHRONO duration_cast<_CHRONO minutes>(_CHRONO abs(_Dur) - hours())},
+              _Secs{_CHRONO duration_cast<_CHRONO seconds>(_CHRONO abs(_Dur) - hours() - minutes())} {
+            // clang-format on
+            if constexpr (treat_as_floating_point_v<typename precision::rep>) {
+                _Sub_secs = _CHRONO abs(_Dur) - hours() - minutes() - seconds();
+            } else {
+                _Sub_secs = _CHRONO duration_cast<precision>(_CHRONO abs(_Dur) - hours() - minutes() - seconds());
+            }
+        }
+
+        _NODISCARD constexpr bool is_negative() const noexcept {
+            return _Is_neg;
+        }
+        _NODISCARD constexpr hours hours() const noexcept {
+            return _Hours;
+        }
+        _NODISCARD constexpr minutes minutes() const noexcept {
+            return _Mins;
+        }
+        _NODISCARD constexpr seconds seconds() const noexcept {
+            return _Secs;
+        }
+        _NODISCARD constexpr precision subseconds() const noexcept {
+            return _Sub_secs;
+        }
+
+        _NODISCARD constexpr explicit operator precision() const noexcept {
+            return to_duration();
+        }
+        _NODISCARD constexpr precision to_duration() const noexcept {
+            const auto _Dur = _Hours + _Mins + _Secs + _Sub_secs;
+            return _Is_neg ? -_Dur : _Dur;
+        }

```

Comment:
> Looking around elsewhere in the codebase for use of static_asserts, I see them predominantly in test.cpps, not product code. I think sticking with the _STL_ASSERT should be okay here, though I'll yield to another maintainer if they disagree.

That is what duration is doing for the numerator 

https://github.com/microsoft/STL/blob/master/stl/inc/chrono#L72

Thought maybe it would be best to repeat that here? :/

(for some reason I can't comment on that comment)